[FLINK-37267][table] Add support for UNNEST WITH ORDINALITY

gustavodemorais opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

The SQL standard specifies a WITH ORDINALITY clause that can be appended to any UNNEST function call, which returns a new row for each element and its position in the data structure being unnested.

Examples:

```sql
– Returns a new row for each element in a constant array and its position in the array
SELECT *
FROM (VALUES('order_1'))
CROSS JOIN UNNEST(ARRAY["shirt", "pants", "hat"])
WITH ORDINALITY AS t(product_name, index)

id       product_name  index
=======  ============  =====
order_1  shirt             1
order_1  pants             2
order_1  hat               3

– Returns a new row for each element and its position in the array assuming a Orders table with an array column `product_names`
SELECT order_id, product_name, product_index
FROM Orders
CROSS JOIN UNNEST(product_names)
WITH ORDINALITY AS t(product_name, product_index)
```

A unnest with ordinality will return each element and the position of the element in the data structure, 1-indexed. The order of the elements for arrays is guaranteed. Since maps and multisets are unordered, the order of the elements is not guaranteed.

## Brief change log

  - Created UnnestRowsWithOrdinalityFunction
  - Refactored code so we have one UnnestRowsFunctionBase for both with and without ordinality
  - Refactored to remove duplicated code: eval and type logic in UnnestRowsFunctionBase. We just `.collect` with or without the index (ordinality) as an extra column depending if with or without ordinality.
 - Call UnnestRowsFunction or UnnestRowsWithOrdinalityFunction in LogicalUnnestRule.java when rewriting tree.
 - For arrays and multisets of rows, we extract the fields and emit one single row with one extra column, following Calcite's implementation.
  - Revamped outdated unnest documentation

## Verifying this change

This change added tests and can be verified as follows:

  -  Added several stream tests, batch tests and plan tests    
   - I'm still testing/checking if we need to add other relevant tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

dawidwys reviewed (changes requested):
Nice work @gustavodemorais 

I put some minor comments.

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java:
```diff
@@ -0,0 +1,198 @@
+                final int multiplier = values.getInt(pos);
+                final Object key = elementGetter.getElementOrNull(keys, pos);
+                for (int i = 0; i < multiplier; i++) {
+                    collector.collect(key, ordinal++);
```
Yes, but since we multisets aren't stored like arrays we have to check the multiplier/how often they appeared instead of just iterating. 

I.e. if we have `element1, element1, element2, element2, element2` as a multiset, we want to return 5 for the last `element2`. Since what is actually stored is {element1: 3, element2: 2}, we first go through the elements (element1 and element2) and then we go through the multiplier (how often they appeared).

dawidwys replied:
ah, true, sorry didn't read it carefully enough 👍

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java:
```diff
@@ -0,0 +1,198 @@
+public abstract class UnnestRowsFunctionBase extends BuiltInSpecializedFunction {
+
+    public UnnestRowsFunctionBase() {
+        super(BuiltInFunctionDefinitions.INTERNAL_UNNEST_ROWS);
```
We can always use the basic internal unnest rows here. The WithOrdinality classes instantiate this base class and override all relevant methods. If I didn't miss anything, I guess we could add one more param and distinguish, but it wouldn't change anything.

gustavodemorais replied:
Sure, done

gustavodemorais replied:
Done

gustavodemorais commented:
Thanks for the review, @dawidwys! Addressed the comments. 

I'm adding tests for arrays with rows since I realized a small difference compared to how calcite implements it. In short, an array with `row(a,b)` unnests to a `row(a, b, ordinality)`. My first assumption was that it unnested to `row(a,b), ordinality`. Making the changes to accommodate this and will push it soon.

gustavodemorais commented:
Apart from addressing the minor comments, I've made the changes to support arrays and multisets of rows, following Calcite's implementation. 

Also added several tests for multiple cases, considering examples and other implementations. The PR should be ready.

gustavodemorais commented:
Apart from addressing the minor comments, I've made the changes to support arrays and multisets of rows, following Calcite's implementation.

After a couple of iterations on the code and refactoring, also added several tests for multiple cases. The PR should be ready.

twalthr commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java:
```diff
@@ -0,0 +1,211 @@
+
+            for (int i = 0; i < fieldCount; i++) {
+                types[i] = rowType.getTypeAt(i);
+                names[i] = "f" + i;
```
use `r =LogicalTypeUtils#toRowType`?
then 
`new RowType(Stream.concat(r.getFields().stream(), Stream.of(new RowField("ordinality", new IntType(false)))`.
we don't want to loose field names and should improve code readability.

gustavodemorais replied:
Oh yes, nice. That makes sense. I was blind to only using RowType.of and didn't consider using new RowType. Since we are have the fields array from `.getFields`, we could do both:

1.
```
return new RowType(
                    false,
                    Stream.concat(
                                    rowType.getFields().stream(),
                                    Stream.of(new RowType.RowField("ordinality", DataTypes.INT().notNull().getLogicalType())))
                            .collect(Collectors.toList()));
```
and
2.
```
return new RowType(
                    false,
                    new ArrayList<>() {{
                        addAll(rowType.getFields());
                        add(new RowType.RowField("ordinality", DataTypes.INT().notNull().getLogicalType()));
                    }}
            );
```

I find 2. a bit easier to read but pushed your suggestion for now (1.) - no strong preference, though. Let me know if you have a preference. According to my findings, the performance should be exactly the same for both O(n + 1).

gustavodemorais replied:
See https://github.com/apache/flink/pull/26113/commits/7f522cfdfe1817dc1c4a0162153ec5e93dd0ec5a

twalthr commented on docs/content/docs/dev/table/sql/queries/joins.md:
```diff
@@ -326,16 +326,55 @@ FROM Orders AS o
+
+-- Returns a new row for each element and its position in the array
+-- assuming a Orders table with an array column `product_names`
+SELECT order_id, product_name, product_index
```
Could we make an example for multisets? It's not clear to me how WITH ORDINALITY behaves with them. The tests seem to drop the index?

gustavodemorais replied:
Sure, adding one example for multisets and one for a map. Multisets are stored as maps, so we go through the keys and emit them x amount of times, x being how often they occurred (their multiplicity). Let me know if the example helped

gustavodemorais replied:
https://github.com/apache/flink/pull/26113/commits/e179c09f3cc980b126bf2a6b4a0b9477518401a6

gustavodemorais commented on docs/content/docs/dev/table/sql/queries/joins.md:
````diff
@@ -326,16 +326,55 @@ FROM Orders AS o
+Unnest returns a new row for each element in the given array, multiset or map. Supports both `CROSS JOIN` and `LEFT JOIN`.
+```sql
+-- Returns a new row for each element in a constant array
+SELECT * FROM (VALUES('order_1')), UNNEST(ARRAY["shirt", "pants", "hat"])
````
Good catch. I rewrote testUnnestWithValuesStream2 from tests so it's more user-friendly and messed up the quotes

gustavodemorais commented on docs/content/docs/dev/table/sql/queries/joins.md:
```diff
@@ -326,16 +326,55 @@ FROM Orders AS o
-FROM Orders CROSS JOIN UNNEST(tags) AS t (tag)
+-- Returns a new row for each element in a constant array and its position in the array
+SELECT * 
+FROM (VALUES('order_1'))
```
> it might be easier to understand with a space after VALUES and a second value
Extended the example with both notes

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunction.java:
```diff
@@ -19,169 +19,89 @@
+        public MapUnnestFunction(
                 SpecializedContext context,
-                LogicalType outputType,
+                LogicalType keyValTypes,
```
These are element types we receive as an input. They might or might not be the same as the output type. For with ordinality we have:

keyValTypes: `row(key, val)`
outputType: `row(key, val, ordinality)`

Since it's still not yet the output type, and the outputType can actually differ, I renamed the params to `elementType` for CollectionUnnest and `keyValTypes` for MapUnnest for both classes. IMO was clearer and closer to reality since. Wdyt?

gustavodemorais commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/UnnestITCase.scala:
```diff
@@ -315,4 +315,279 @@ class UnnestITCase extends BatchTestBase {
+    registerCollection(
+      "T",
+      data,
+      new RowTypeInfo(Types.INT, Types.PRIMITIVE_ARRAY(Types.INT)),
```
I don't think so. They all take TypeInformation[T] as a param unfortunately

gustavodemorais commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/UnnestITCase.scala:
```diff
@@ -379,4 +380,486 @@ class UnnestITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mo
+    env.execute()
+
+    val expected = List("1,12,45.6,1", "2,12,45.612,1", "2,13,41.6,2")
+    assertThat(sink.getRetractResults.sorted).isEqualTo(expected.sorted)
```
See https://github.com/apache/flink/pull/26113#discussion_r1957125379

gustavodemorais commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/UnnestITCase.scala:
```diff
@@ -315,4 +315,279 @@ class UnnestITCase extends BatchTestBase {
+    val data = List(
+      row(
+        1, {
+          val map = new java.util.HashMap[Row, Row]()
```
Ops, done. I think this was the only place where I didn't use asJava. See https://github.com/apache/flink/pull/26113/commits/8d964daf07e8f71637f5d9fb9371342e582a6bd5

gustavodemorais commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/UnnestITCase.scala:
```diff
@@ -379,4 +380,486 @@ class UnnestITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mo
+      (3, (14, "45.2136")),
+      (3, (18, "42.6")))
+    val t = StreamingEnvUtil.fromCollection(env, data).toTable(tEnv, 'a, 'b)
+    tEnv.createTemporaryView("T", t)
```
See https://github.com/apache/flink/commit/e54bf99dd04c4f8139eeda71472746bb523d6081

gustavodemorais commented:
Hey @twalthr, thanks for the review! I've addressed all the points. An important shout is that I'll definitely be looking into using semantic tests for future changes.

Currently, we already have 3 different files for all unnest testing, and I was hesitant at first to create a new file and spread them even more and changing old tests in this PR. Anyway, I've put some time into refactoring unnest stream tests within the file, so it's easier to read and eventually easier to migrate :)

twalthr commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/UnnestITCase.scala:
```diff
@@ -312,71 +306,509 @@ class UnnestITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mo
+                   |""".stripMargin,
+      expectedResults = List("1,a,10,1", "1,b,11,2", "2,c,20,2", "2,d,21,1"),
+      isRetract = false,
+      parallelism = 1,
```
Is this parallelism important for the behavior of the test? if not, we can run all tests with the same value and remove this parameter from `assertUnnest()`

twalthr commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java:
```diff
@@ -0,0 +1,219 @@
+            return RowType.of(
+                    false,
+                    new LogicalType[] {baseType, DataTypes.INT().notNull().getLogicalType()},
+                    new String[] {"f0", "ordinality"});
```
For PTF [we changed the default](https://github.com/apache/flink/blob/master/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/SystemTypeInference.java#L272) from `f0` to `EXPR$0`. We should do the same here.

twalthr commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsFunctionBase.java:
```diff
@@ -0,0 +1,219 @@
+                                    rowType.getFields().stream(),
+                                    Stream.of(
+                                            new RowType.RowField(
+                                                    "ordinality",
```
```suggestion
                                                    "ORDINALITY",
```

twalthr reviewed (commented):
Thanks for the update @gustavodemorais. I added my last set of feedback. Should be good in the next iteration :-)

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/UnnestRowsWithOrdinalityFunction.java:
```diff
@@ -0,0 +1,149 @@
+            if (element instanceof RowData) {
+                RowData innerRow = (RowData) element;
+                int arity = innerRow.getArity();
+                GenericRowData outRow = new GenericRowData(arity + 1);
```
Great tip, thanks! See https://github.com/apache/flink/pull/26113/commits/5c72d20acdb392f8f82d1b0b9d813206064d6774

gustavodemorais commented:
Thanks for the review, @twalthr! Please take a look at the updated code

twalthr reviewed (approved):
Thanks for the update @gustavodemorais. LGTM.
